draft-jmiller-json-proof-algorithms.md

```diff
@@ -404,26 +404,57 @@ Each payload is serialized using UTF-8 and encoded with base64url into an octet
-WIP: presentation_proof = [issuer_signature, revealed_key_0, hidden_mac_1, hidden_mac_2, revealed_key_3]
+These two signatures are then followed by a MAC value for each payload.  The MAC values used will depend on if that payload has been disclosed or is hidden.  Disclosed payloads will include the MAC key input, and hidden payloads will include only their final MAC value.
 
-Using the ES256 algorithm with a signature of 64 octets and for a JWP with five payloads using HMAC-SHA256 the total presentation proof value length would be `64 + (5 * 32) = 224` octets.
+`presentation_proof = [presentation_signature, issuer_signature. disclosed_key_0, hidden_mac_1, hidden_mac_2, ... disclosed_key_n]`
 
-### Verification
+The size of this value will depend on the underlying cryptographich algorithms.  For example, `MAC-H256` uses the `ES256` JWS with a decoded signature of 64 octets, and for a JWP with five payloads using `HMAC-SHA256` the total presentation proof value length would be `64 + 64 + (5 * 32) = 288` octets.
+
+### Verifier Setup
+
+In order to verify that the presentation was protected from replay attacks, the verifier must be able to validate the presentation protected header.  This involves the following steps:
+
+1. JSON parse the presentation header
+2. Validate the contained `nonce` claim
+3. JSON parse the issuer header
+4. Validate the contained `pjwk` claim
+5. Create a JWS using the correct fixed header with `alg` value and the presentation header as the body
+6. Remove the `presentation_signature` from the beginning of the `presentation_proof` octet array
+7. Validate the JWS using the JWK from the `pjwk` claim and the `presentation_signature` value
+
+Next, the verifier must validate all of the disclosed payloads using the following steps:
 
-WIP: generate a MAC for the revealed payloads w/ the included keys, generate array of final MACs and validate issuer signature, validate presentation signature.
+1. JSON parse the issuer header
+2. Resolve the `kid` using a trusted mechanism to obtain the correct issuer JWK
+3. Remove the `issuer_signature` from the beginning of the remaining `presentation_proof` octet array (after the `presentation_signature` was removed)
+4. Perform the MAC on the presented `issuer_header` value using the "issuer_header" value as the input key
+5. Store the resulting value as the first entry in a new `jws_payload` octet array
+6. Iterate on each presented payload (disclosed or hidden)
+   1. Extract the next hash value from the remaining `presentation_proof` octet array
+   2. If the payload was disclosed: perform a MAC using the given hash value as the input key and append the result to the `jws_payload` octet array
+   3. If the payload was hidden: append the given hash value to the `jws_payload` octet array
+7. Create a JWS using the correct fixed header with `alg` value and the generated `jws_payload` value as the payload
```

Comment:
"the correct fixed header with" ? not sure abut this language, could be clear.

Reply:
Good eye! I literally felt the hesitation as I was typing it that I needed to explain that better, but I pushed on just to get the larger body all written first.

I'll circle back and make another pass, but I don't expect it to be perfect in this PR as there's a few larger editorial/cleanup ones that will be fast following it.